File: modules/vescape-core/ios/protocol/VescGattClient.swift
import CoreBluetooth
import Foundation

/// Connection + scan lifecycle callbacks. All fire on the main queue (the central runs there),
/// so the coordinator mutates Board Session state without extra hopping. Mirrors the Android
/// `VescGattListener` phase surface.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/protocol/VescGattClient.kt
internal protocol VescGattListener: AnyObject {
  func onDeviceDiscovered(id: String, name: String, rssi: Int, serviceUUIDs: [String])
  func onScanFailure(_ message: String)
  func onGattConnected()
  func onGattSubscribing()
  func onGattReady()
  func onGattDisconnected(intentional: Bool, message: String)
  func onGattFailure(code: String, message: String)
  func onGattFrameChunk(_ chunk: [UInt8])
  /// CoreBluetooth handed the session central's state back after the app was relaunched into the
  /// background (ADR 0034). `peripheralIds` are the peripherals iOS restored — empty when the link
  /// died while the process was dead, which is a resume through the normal reconnect path.
  ///
  /// @platform-diff No Android peer: its `CoreForegroundService` keeps the process alive, so there
  /// is nothing to restore.
  func onGattRestored(peripheralIds: [String])
}

/// Transport seam under `BoardSessionController` (ADR 0024): everything the controller calls on
/// the board link beyond the `VescGattListener` callbacks. The real `VescGattClient` speaks
/// CoreBluetooth; the dev-mode `ReplayTransport` plays a Debug Recording through the same surface.
/// `supportsReconnect` gates the reconnect loop: a replay ending is terminal, not recoverable.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/protocol/VescGattClient.kt `SessionTransport`
internal protocol SessionTransport: AnyObject {
  var supportsReconnect: Bool { get }
  func connect(peripheralId: String)
  func disconnect()
  func reconnect()
  func startReconnectScan()
  func stopReconnectScan()
  @discardableResult
  func sendPayload(_ payload: [UInt8]) -> Bool
}

/// CoreBluetooth wrapper around a single VESC board connection plus BLE scanning. Owns one
/// `CBCentralManager` shared by scan and connect; the coordinator drives phases through the
/// listener. Deliberately dumb: it exposes a persistent `reconnect()` (re-issuing CoreBluetooth's
/// self-retrying `connect`) and a supplemental rescan, but the reconnect *policy* — when to give
/// up, phase transitions, session identity — lives one layer up.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/protocol/VescGattClient.kt
internal final class VescGattClient: NSObject, SessionTransport {
  var supportsReconnect: Bool { true }
  /// Active raw debug Session Recorder resolver, set by the session controller. Records `tx`
  /// chunks at the write site (the peer of Android taping `tx` in its `VescGattClient`).
  /// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/protocol/VescGattClient.kt `recorder`
  var recorder: (() -> SessionRecorder?)?
  private weak var listener: VescGattListener?
  /// Stable across app versions on purpose: iOS keys the preserved central state on this string, so
  /// changing it throws away every in-flight restoration. Only the Board Session client passes it;
  /// the Board Probe's client (`BoardTransportDetector`) stays bare — probing never needs
  /// resurrection, and two centrals may not share one restore identifier.
  static let sessionRestoreIdentifier = "com.vescape.core.session-central"
  private let restoreIdentifier: String?
  private lazy var central: CBCentralManager = {
    guard let restoreIdentifier else { return CBCentralManager(delegate: self, queue: nil) }
    return CBCentralManager(
      delegate: self,
      queue: nil,
      options: [CBCentralManagerOptionRestoreIdentifierKey: restoreIdentifier]
    )
  }()
  /// Peripherals handed back by `willRestoreState`, held until the coordinator adopts one.
  private var restoredPeripherals: [CBPeripheral] = []
  /// Adoption requested before the central reported `.poweredOn` (restoration delivers
  /// `willRestoreState` first), completed from `centralManagerDidUpdateState`.
  private var pendingRestoreAdoptId: UUID?

  private var peripheral: CBPeripheral?
  private var txChar: CBCharacteristic?
  private var writeType: CBCharacteristicWriteType = .withoutResponse
  private var pendingNotifyEnables = 0
  private var readyResolved = false
  private var intentionalDisconnect = false

  /// True while emitting discovery devices to the listener (the `scan()` JS API).
  private var isDiscoveryScanning = false
  /// Set while scanning to locate a specific board to connect to.
  private var connectTargetId: UUID?

  /// Deferred until the central reports `.poweredOn`.
  private var pendingDiscoveryScan = false
  private var pendingConnectId: UUID?
  /// A persistent reconnect queued while the central was not yet `.poweredOn`.
  private var pendingReconnect = false
  /// Target of the active connect session, retained across reconnects so a board that never
  /// connected this session (e.g. powered off at launch) can still be re-resolved. Cleared only by
  /// an intentional `disconnect()`.
  private var lastConnectId: UUID?
  /// True while a reconnect scans for its target without a retained peripheral, so the supplemental
  /// rescan windows don't tear that scan down during their idle gaps.
  private var reconnectTargetScan = false

  /// `restoreIdentifier` opts this client's central into CoreBluetooth state restoration (ADR 0034).
  init(listener: VescGattListener, restoreIdentifier: String? = nil) {
    self.listener = listener
    self.restoreIdentifier = restoreIdentifier
    super.init()
    _ = central // Kick off state updates so poweredOn arrives before first use.
  }

  // MARK: - Scan (JS `scan()` API)

  func startScan() {
    isDiscoveryScanning = true
    guard central.state == .poweredOn else {
      pendingDiscoveryScan = true
      return
    }
    beginScan()
  }

  func stopScan() {
    isDiscoveryScanning = false
    pendingDiscoveryScan = false
    // Keep scanning if a connect is still hunting for its target peripheral.
    if connectTargetId == nil {
      central.stopScan()
    }
  }

  // MARK: - Connect

  func connect(peripheralId: String) {
    guard let uuid = UUID(uuidString: peripheralId) else {
      listener?.onGattFailure(code: "INVALID_DEVICE", message: "Malformed BLE id: \(peripheralId)")
      return
    }
    // A lingering peripheral from a previous attempt keeps delivering callbacks; tear it down.
    clear(markIntentional: true)
    intentionalDisconnect = false
    readyResolved = false
    lastConnectId = uuid

    guard central.state == .poweredOn else {
      pendingConnectId = uuid
      return
    }
    connectResolved(uuid)
  }

  private func connectResolved(_ uuid: UUID) {
    if let known = central.retrievePeripherals(withIdentifiers: [uuid]).first {
      connectPeripheral(known)
      return
    }
    if let live = central.retrieveConnectedPeripherals(withServices: [VescGattUUIDs.service]).first(where: {
      $0.identifier == uuid
    }) {
      connectPeripheral(live)
      return
    }
    // Not yet seen this launch: scan until the board advertises, then connect.
    connectTargetId = uuid
    beginScan()
  }

  private func connectPeripheral(_ peripheral: CBPeripheral) {
    reconnectTargetScan = false
    self.peripheral = peripheral
    peripheral.delegate = self
    central.connect(peripheral, options: nil)
  }

  func disconnect() {
    lastConnectId = nil
    restoredPeripherals = []
    pendingRestoreAdoptId = nil
    clear(markIntentional: true)
  }

  // MARK: - State restoration (#378)

  /// Adopt a peripheral CoreBluetooth restored into this launch as the session's link. The GATT
  /// subscriptions survived the process death, but this object graph did not: the delegate is
  /// re-set and services re-discovered so `txChar` and the frame callbacks exist again.
  ///
  /// Returns false when the id is not among the restored peripherals, which routes the coordinator
  /// to a normal persistent connect instead.
  @discardableResult
  func adoptRestored(peripheralId: String) -> Bool {
    guard
      let uuid = UUID(uuidString: peripheralId),
      let restored = restoredPeripherals.first(where: { $0.identifier == uuid })
    else { return false }
    restoredPeripherals = []
    intentionalDisconnect = false
    readyResolved = false
    txChar = nil
    pendingNotifyEnables = 0
    lastConnectId = uuid
    peripheral = restored
    restored.delegate = self
    guard central.state == .poweredOn else {
      // Restoration delivers `willRestoreState` before the first state update; finish once the
      // radio reports in, or CoreBluetooth silently drops the discovery/connect.
      pendingRestoreAdoptId = uuid
      return true
    }
    resumeAdopted(restored)
    return true
  }

  private func resumeAdopted(_ peripheral: CBPeripheral) {
    guard peripheral === self.peripheral else { return }
    if peripheral.state == .connected {
      listener?.onGattConnected()
      peripheral.discoverServices([VescGattUUIDs.service])
    } else {
      // Restored but dropped since: persistent connect keeps retrying on its own.
      central.connect(peripheral, options: nil)
    }
  }

  // MARK: - Reconnect (#58)

  /// Re-issue a persistent connect on the peripheral retained across an unintentional disconnect.
  /// `connect(_:options:)` is self-retrying — CoreBluetooth keeps trying until it succeeds or the
  /// connection is cancelled, even waking the app from suspension — so there is nothing to poll.
  ///
  /// @platform-diff Android has no gatt-level peer: its `connectGatt(autoConnect = false)` is
  /// one-shot and the `ReconnectScheduler` drives retries with a fresh connect each attempt. iOS
  /// leans on CoreBluetooth's built-in persistent connect instead (see `BoardSessionController`).
  func reconnect() {
    intentionalDisconnect = false
    readyResolved = false
    txChar = nil
    pendingNotifyEnables = 0
    guard central.state == .poweredOn else {
      pendingReconnect = true
      return
    }
    guard let peripheral else {
      reconnectViaTargetResolve()
      return
    }
    central.connect(peripheral, options: nil)
  }

  /// Reconnect variant for a board that never yielded a retained peripheral this session (powered
  /// off since launch). Re-resolves from the stored target so a persistent connect or target scan
  /// keeps retrying instead of dead-ending. The peripheral-backed `reconnect()` is preferred when
  /// one exists; this is the fallback the coordinator's reconnect loop leans on otherwise.
  private func reconnectViaTargetResolve() {
    guard let target = lastConnectId else {
      listener?.onGattFailure(code: "RECONNECT_FAILED", message: "No peripheral to reconnect")
      return
    }
    reconnectTargetScan = true
    connectResolved(target)
  }

  /// Open a supplemental scan window that reconnects the moment the retained peripheral advertises,
  /// accelerating rediscovery on top of the persistent connect's passive retry.
  func startReconnectScan() {
    guard let peripheral, central.state == .poweredOn else { return }
    connectTargetId = peripheral.identifier
    beginScan()
  }

  /// Close the supplemental scan window (persistent connect keeps running). A live JS `scan()`
  /// keeps the radio scanning.
  func stopReconnectScan() {
    // Keep scanning if a live JS `scan()` needs it, or if a peripheral-less reconnect is still
    // hunting for its target (tearing that down would abort the retry).
    if !isDiscoveryScanning && !reconnectTargetScan {
      central.stopScan()
    }
  }

  func sendPayload(_ payload: [UInt8]) -> Bool {
    guard let peripheral, let txChar else { return false }
    let bytes = VescPacketCodec.encode(payload)
    peripheral.writeValue(Data(bytes), for: txChar, type: writeType)
    recorder?()?.recordChunk(direction: "tx", bytes: bytes)
    return true
  }

  // MARK: - Teardown

  private func clear(markIntentional: Bool) {
    connectTargetId = nil
    reconnectTargetScan = false
    pendingReconnect = false
    if !isDiscoveryScanning {
      central.stopScan()
    }
    if let peripheral {
      if markIntentional { intentionalDisconnect = true }
      central.cancelPeripheralConnection(peripheral)
    }
    peripheral = nil
    txChar = nil
    pendingNotifyEnables = 0
  }

  private func beginScan() {
    // Scan unfiltered, mirroring Android's null-filter scan. VESC boards (Nordic UART BLE
    // modules) don't advertise the NUS service UUID in the advertisement packet — it only
    // appears in the GATT table after connecting — so `withServices: [NUS]` never surfaces
    // them. The board is identified by name/id at the UI layer, not by advertised service.
    central.scanForPeripherals(
      withServices: nil,
      options: [CBCentralManagerScanOptionAllowDuplicatesKey: true]
    )
  }

  private func resolveReady() {
    guard !readyResolved else { return }
    readyResolved = true
    listener?.onGattReady()
  }
}

// MARK: - CBCentralManagerDelegate

extension VescGattClient: CBCentralManagerDelegate {
  func centralManagerDidUpdateState(_ central: CBCentralManager) {
    switch central.state {
    case .poweredOn:
      if pendingDiscoveryScan {
        pendingDiscoveryScan = false
        beginScan()
      }
      if let uuid = pendingConnectId {
        pendingConnectId = nil
        connectResolved(uuid)
      }
      if pendingReconnect {
        pendingReconnect = false
        reconnect()
      }
      if let adoptId = pendingRestoreAdoptId {
        pendingRestoreAdoptId = nil
        if let peripheral, peripheral.identifier == adoptId {
          resumeAdopted(peripheral)
        }
      }
    case .poweredOff:
      if isDiscoveryScanning || pendingDiscoveryScan {
        listener?.onScanFailure("Bluetooth is off")
      }
      // A restored session waiting to adopt its peripheral is not a failed connect: failing it here
      // would tear the session down and drop the durable resume marker, so a later power-on could
      // never finish the adoption. Radio off is a pause; the adoption stays deferred (ADR 0034).
      if pendingRestoreAdoptId != nil { return }
      if peripheral != nil || connectTargetId != nil || pendingConnectId != nil {
        listener?.onGattFailure(code: "BLE_OFF", message: "Bluetooth is off")
      }
    case .unauthorized:
      listener?.onScanFailure("Bluetooth permission denied")
    default:
      break
    }
  }

  /// iOS relaunched the app (headlessly, on a board notification) and is handing the session
  /// central's preserved state back. Peripherals are retained and their delegate re-set here; the
  /// decision of what to do with them — rebuild the Board Session, or fall back to a normal
  /// reconnect when none came back — belongs to the coordinator (ADR 0034).
  func centralManager(_ central: CBCentralManager, willRestoreState dict: [String: Any]) {
    let peripherals = (dict[CBCentralManagerRestoredStatePeripheralsKey] as? [CBPeripheral]) ?? []
    for peripheral in peripherals {
      peripheral.delegate = self
    }
    restoredPeripherals = peripherals
    listener?.onGattRestored(peripheralIds: peripherals.map { $0.identifier.uuidString })
  }

  func centralManager(
    _ central: CBCentralManager,
    didDiscover peripheral: CBPeripheral,
    advertisementData: [String: Any],
    rssi RSSI: NSNumber
  ) {
    if let target = connectTargetId, peripheral.identifier == target {
      connectTargetId = nil
      if !isDiscoveryScanning { central.stopScan() }
      connectPeripheral(peripheral)
      return
    }
    guard isDiscoveryScanning else { return }
    let name = (advertisementData[CBAdvertisementDataLocalNameKey] as? String)
      ?? peripheral.name
      ?? ""
    let serviceUUIDs = (advertisementData[CBAdvertisementDataServiceUUIDsKey] as? [CBUUID])?
      .map { $0.uuidString.lowercased() } ?? []
    listener?.onDeviceDiscovered(
      id: peripheral.identifier.uuidString,
      name: name,
      rssi: RSSI.intValue,
      serviceUUIDs: serviceUUIDs
    )
  }

  func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
    guard peripheral === self.peripheral else { return }
    listener?.onGattConnected()
    peripheral.discoverServices([VescGattUUIDs.service])
  }

  func centralManager(
    _ central: CBCentralManager,
    didFailToConnect peripheral: CBPeripheral,
    error: Error?
  ) {
    guard peripheral === self.peripheral else { return }
    listener?.onGattFailure(code: "CONNECT_FAILED", message: error?.localizedDescription ?? "Connect failed")
  }

  func centralManager(
    _ central: CBCentralManager,
    didDisconnectPeripheral peripheral: CBPeripheral,
    error: Error?
  ) {
    guard peripheral === self.peripheral else { return }
    let wasIntentional = intentionalDisconnect
    intentionalDisconnect = false
    txChar = nil
    pendingNotifyEnables = 0
    // Keep the peripheral reference on an unexpected drop so the coordinator can hand it back to a
    // persistent `reconnect()`; only an intentional teardown releases it (via `clear`).
    if wasIntentional {
      self.peripheral = nil
    }
    listener?.onGattDisconnected(
      intentional: wasIntentional,
      message: error?.localizedDescription ?? "Board disconnected"
    )
  }
}

// MARK: - CBPeripheralDelegate

extension VescGattClient: CBPeripheralDelegate {
  func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
    guard peripheral === self.peripheral else { return }
    listener?.onGattSubscribing()
    if let error {
      listener?.onGattFailure(code: "DISCOVERY_FAILED", message: error.localizedDescription)
      return
    }
    guard let service = peripheral.services?.first(where: { $0.uuid == VescGattUUIDs.service }) else {
      listener?.onGattFailure(code: "NO_CHAR", message: "NUS service not found")
      return
    }
    peripheral.discoverCharacteristics([VescGattUUIDs.tx, VescGattUUIDs.rx], for: service)
  }

  func peripheral(
    _ peripheral: CBPeripheral,
    didDiscoverCharacteristicsFor service: CBService,
    error: Error?
  ) {
    guard peripheral === self.peripheral else { return }
    if let error {
      listener?.onGattFailure(code: "DISCOVERY_FAILED", message: error.localizedDescription)
      return
    }
    let chars = service.characteristics ?? []
    guard
      let tx = chars.first(where: { $0.uuid == VescGattUUIDs.tx }),
      let rx = chars.first(where: { $0.uuid == VescGattUUIDs.rx })
    else {
      listener?.onGattFailure(code: "NO_CHAR", message: "NUS characteristics not found")
      return
    }
    txChar = tx
    writeType = tx.properties.contains(.write) ? .withResponse : .withoutResponse

    // Subscribe to whichever of tx/rx actually notify; the board streams telemetry there.
    let notifiers = [rx, tx].filter { $0.properties.contains(.notify) || $0.properties.contains(.indicate) }
    guard !notifiers.isEmpty else {
      listener?.onGattFailure(code: "NO_CHAR", message: "No notifying NUS characteristic")
      return
    }
    pendingNotifyEnables = notifiers.count
    for characteristic in notifiers {
      peripheral.setNotifyValue(true, for: characteristic)
    }
    // Some boards never ack the subscribe; resolve after a grace period so connect never hangs.
    DispatchQueue.main.asyncAfter(deadline: .now() + 4) { [weak self] in
      self?.resolveReady()
    }
  }

  func peripheral(
    _ peripheral: CBPeripheral,
    didUpdateNotificationStateFor characteristic: CBCharacteristic,
    error: Error?
  ) {
    guard peripheral === self.peripheral else { return }
    pendingNotifyEnables = max(0, pendingNotifyEnables - 1)
    if pendingNotifyEnables == 0 {
      resolveReady()
    }
  }

  func peripheral(
    _ peripheral: CBPeripheral,
    didUpdateValueFor characteristic: CBCharacteristic,
    error: Error?
  ) {
    guard peripheral === self.peripheral else { return }
    guard characteristic.uuid == VescGattUUIDs.rx || characteristic.uuid == VescGattUUIDs.tx else { return }
    guard let value = characteristic.value else { return }
    listener?.onGattFrameChunk([UInt8](value))
  }
}

/// NUS UUIDs as `CBUUID`, sourced from the shared `VescUartUUIDs` so iOS keeps one truth.
internal enum VescGattUUIDs {
  static let service = CBUUID(nsuuid: VescUartUUIDs.service)
  static let tx = CBUUID(nsuuid: VescUartUUIDs.tx)
  static let rx = CBUUID(nsuuid: VescUartUUIDs.rx)
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/protocol/VescGattClient.kt
package expo.modules.vescapecore.protocol

import expo.modules.vescapecore.recording.SessionRecorder
import expo.modules.vescapecore.service.VESC_SESSION_TAG

import android.annotation.SuppressLint
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothProfile
import android.bluetooth.BluetoothStatusCodes
import android.content.Context
import android.os.Build
import android.os.Handler
import android.util.Log
import java.util.UUID

private val NUS_SERVICE_UUID = UUID.fromString("6e400001-b5a3-f393-e0a9-e50e24dcca9e")
private val NUS_TX_UUID = UUID.fromString("6e400002-b5a3-f393-e0a9-e50e24dcca9e")
private val NUS_RX_UUID = UUID.fromString("6e400003-b5a3-f393-e0a9-e50e24dcca9e")
private val CCCD_UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

internal interface VescGattListener {
    fun onGattConnected()
    fun onGattSubscribing()
    fun onGattDisconnected(status: Int, intentional: Boolean)
    fun onGattReady()
    fun onGattFailure(code: String, message: String)
    fun onGattFrameChunk(chunk: ByteArray)
}

/**
 * Transport seam under [BoardSessionController] (ADR 0024): everything the controller calls on the
 * board link beyond the [VescGattListener] callbacks. The real [VescGattClient] speaks GATT; the
 * dev-mode ReplayTransport plays a Debug Recording through the same surface.
 * @parity /modules/vescape-core/ios/protocol/VescGattClient.swift `SessionTransport`
 */
internal interface SessionTransport {
    fun connect(deviceId: String)
    fun sendPayload(payload: ByteArray): Boolean
    fun sendRemoteInput(payload: ByteArray, urgent: Boolean = false): Boolean
    fun clear(markIntentional: Boolean = true)
}

// @parity /modules/vescape-core/ios/protocol/VescGattClient.swift
@SuppressLint("MissingPermission")
internal class VescGattClient(
    private val context: Context,
    private val handler: Handler,
    private val recorder: () -> SessionRecorder?,
    private val listener: VescGattListener,
    private val dispatchListener: ((() -> Unit) -> Unit) = { it() },
) : SessionTransport {
    private var gatt: BluetoothGatt? = null
    private var txChar: BluetoothGattCharacteristic? = null
    private var pendingCccdWrites = 0
    private var cccdTimeout: Runnable? = null
    private var writeRetry: Runnable? = null
    private var intentionalDisconnect = false
    private val writeQueue = VescWriteQueue()

    override fun connect(deviceId: String) {
        val adapter = (context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager).adapter
        val device = adapter.getRemoteDevice(deviceId)
        Log.d(VESC_SESSION_TAG, "gatt connect request device=${device.address}")
        // A lingering gatt from a previous attempt keeps delivering callbacks on the
        // shared callback object and would race this connection; tear it down first.
        if (gatt != null) clear(markIntentional = true)
        // Each connection starts unintentional; the teardown flag belongs to the gatt
        // we just cleared, not to the new one.
        intentionalDisconnect = false
        gatt = device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
    }

    override fun sendPayload(payload: ByteArray): Boolean {
        if (gatt == null || txChar == null) return false
        writeQueue.enqueueNormal(VescPacketCodec.encode(payload))
        return drainWriteQueue()
    }

    /**
     * Enqueue transient remote input (tilt or Board Move). Only latest unsent value survives, so an
     * emergency neutral command cannot sit behind stale tilt commands.
     */
    override fun sendRemoteInput(payload: ByteArray, urgent: Boolean): Boolean {
        if (gatt == null || txChar == null) return false
        writeQueue.replaceRemoteInput(VescPacketCodec.encode(payload), urgent)
        return drainWriteQueue()
    }

    override fun clear(markIntentional: Boolean) {
        try {
            cancelCccdTimeout()
            writeRetry?.let { handler.removeCallbacks(it) }
            writeRetry = null
            writeQueue.clear()
            if (markIntentional && gatt != null) intentionalDisconnect = true
            gatt?.disconnect()
            gatt?.close()
        } catch (e: Exception) {
            Log.w(VESC_SESSION_TAG, "GATT cleanup failed: ${e.message}")
        }
        gatt = null
        txChar = null
    }

    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            Log.d(VESC_SESSION_TAG, "onConnectionStateChange status=$status newState=$newState")
            // Late callback from a previous (already-replaced/cleared) connection. Close it
            // and leave the current session's state untouched — otherwise a stale disconnect
            // would clobber the live gatt and freeze telemetry.
            if (gatt !== this@VescGattClient.gatt) {
                try { gatt.close() } catch (e: Exception) { Log.w(VESC_SESSION_TAG, "stale gatt close failed: ${e.message}") }
                return
            }
            recorder()?.recordState("gatt:$newState", mapOf("status" to status))
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    dispatchListener { listener.onGattConnected() }
                    val requested = gatt.requestMtu(517)
                    Log.d(VESC_SESSION_TAG, "gatt requestMtu requested=$requested")
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    val wasIntentional = intentionalDisconnect
                    clear(markIntentional = false)
                    if (wasIntentional) intentionalDisconnect = false
                    dispatchListener { listener.onGattDisconnected(status, wasIntentional) }
                }
            }
        }

        override fun onMtuChanged(gatt: BluetoothGatt, mtu: Int, status: Int) {
            if (gatt !== this@VescGattClient.gatt) return
            Log.d(VESC_SESSION_TAG, "onMtuChanged mtu=$mtu status=$status")
            val discoveryStarted = gatt.discoverServices()
            Log.d(VESC_SESSION_TAG, "gatt discoverServices started=$discoveryStarted")
            if (!discoveryStarted) {
                dispatchListener { listener.onGattFailure("DISCOVERY_FAILED", "Could not start service discovery") }
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            if (gatt !== this@VescGattClient.gatt) return
            Log.d(VESC_SESSION_TAG, "onServicesDiscovered status=$status")
            dispatchListener { listener.onGattSubscribing() }
            if (status != BluetoothGatt.GATT_SUCCESS) {
                dispatchListener { listener.onGattFailure("DISCOVERY_FAILED", "Service discovery failed status=$status") }
                return
            }
            val service = gatt.getService(NUS_SERVICE_UUID)
            val tx = service?.getCharacteristic(NUS_TX_UUID)
            val rx = service?.getCharacteristic(NUS_RX_UUID)
            if (service == null || tx == null || rx == null) {
                dispatchListener { listener.onGattFailure("NO_CHAR", "NUS service/characteristics not found") }
                return
            }
            txChar = tx
            val highPriority = gatt.requestConnectionPriority(BluetoothGatt.CONNECTION_PRIORITY_HIGH)
            Log.d(VESC_SESSION_TAG, "gatt requestConnectionPriority high=$highPriority")
            val rxNotify = gatt.setCharacteristicNotification(rx, true)
            val txNotify = gatt.setCharacteristicNotification(tx, true)
            Log.d(VESC_SESSION_TAG, "gatt set notifications rx=$rxNotify tx=$txNotify")

            val rxCccd = rx.getDescriptor(CCCD_UUID)
            if (rxCccd == null) {
                dispatchListener { listener.onGattReady() }
                return
            }
            pendingCccdWrites = 1
            if (tx.getDescriptor(CCCD_UUID) != null) pendingCccdWrites = 2
            Log.d(VESC_SESSION_TAG, "gatt cccd writes pending=$pendingCccdWrites")
            writeCccd(gatt, rxCccd)

            cccdTimeout = Runnable {
                Log.w(VESC_SESSION_TAG, "CCCD ack timeout, resolving connect pending=$pendingCccdWrites")
                dispatchListener { listener.onGattReady() }
            }
            handler.postDelayed(cccdTimeout!!, 4000)
        }

        override fun onDescriptorWrite(gatt: BluetoothGatt, descriptor: BluetoothGattDescriptor, status: Int) {
            if (gatt !== this@VescGattClient.gatt) return
            if (descriptor.uuid != CCCD_UUID) return
            Log.d(VESC_SESSION_TAG, "onDescriptorWrite status=$status pendingBefore=$pendingCccdWrites")
            pendingCccdWrites--
            if (pendingCccdWrites > 0) {
                val txCccd = gatt.getService(NUS_SERVICE_UUID)
                    ?.getCharacteristic(NUS_TX_UUID)
                    ?.getDescriptor(CCCD_UUID)
                if (txCccd != null) {
                    writeCccd(gatt, txCccd)
                    return
                }
            }
            cancelCccdTimeout()
            dispatchListener { listener.onGattReady() }
        }

        /** API 33+ signature. Below 33 the framework calls the deprecated 2-param peer instead. */
        override fun onCharacteristicChanged(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            value: ByteArray,
        ) {
            handleNotification(gatt, characteristic, value)
        }

        @Deprecated("Deprecated in Java")
        @Suppress("DEPRECATION")
        override fun onCharacteristicChanged(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
        ) {
            handleNotification(gatt, characteristic, characteristic.value ?: return)
        }

        private fun handleNotification(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            value: ByteArray,
        ) {
            if (gatt !== this@VescGattClient.gatt) return
            if (characteristic.uuid == NUS_RX_UUID || characteristic.uuid == NUS_TX_UUID) {
                val chunk = value.copyOf()
                dispatchListener { listener.onGattFrameChunk(chunk) }
            }
        }

        override fun onCharacteristicWrite(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            status: Int,
        ) {
            if (gatt !== this@VescGattClient.gatt || characteristic.uuid != NUS_TX_UUID) return
            val completed = writeQueue.completeInFlight()
            val bytes = completed?.bytes
            if (status != BluetoothGatt.GATT_SUCCESS) {
                Log.w(VESC_SESSION_TAG, "gatt write callback failed status=$status bytes=${bytes?.size}")
            }
            drainWriteQueue()
        }
    }

    /** Android permits one characteristic write at a time. Serializing all writes
     * prevents telemetry polling and held remote controls from dropping each other. */
    private fun drainWriteQueue(): Boolean {
        val g = gatt ?: return false
        val tx = txChar ?: return false
        val write = writeQueue.startNext() ?: return true
        val bytes = write.bytes
        writeRetry?.let { handler.removeCallbacks(it) }
        writeRetry = null
        val ok = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            g.writeCharacteristic(
                tx,
                bytes,
                BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT,
            ) == BluetoothStatusCodes.SUCCESS
        } else {
            @Suppress("DEPRECATION")
            run {
                tx.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
                tx.value = bytes
                g.writeCharacteristic(tx)
            }
        }
        if (!ok) {
            writeQueue.retryInFlight()
            Log.w(VESC_SESSION_TAG, "gatt writeCharacteristic failed bytes=${bytes.size}; retrying")
            val retry = Runnable { drainWriteQueue() }
            writeRetry = retry
            handler.postDelayed(retry, 25L)
            return false
        }
        recorder()?.recordChunk("tx", bytes)
        return ok
    }

    private fun writeCccd(gatt: BluetoothGatt, descriptor: BluetoothGattDescriptor) {
        val ok = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            gatt.writeDescriptor(descriptor, BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE) ==
                BluetoothStatusCodes.SUCCESS
        } else {
            @Suppress("DEPRECATION")
            run {
                descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                gatt.writeDescriptor(descriptor)
            }
        }
        Log.d(VESC_SESSION_TAG, "gatt writeCccd started=$ok")
    }

    private fun cancelCccdTimeout() {
        cccdTimeout?.let { handler.removeCallbacks(it) }
        cccdTimeout = null
    }
}
